src/property-filter/index.tsx

```diff
@@ -158,46 +159,58 @@ const PropertyFilter = React.forwardRef(
-        getValueFormatter: operator => (operator ? extendedOperators.get(operator)?.format ?? null : null),
-        getValueFormRenderer: operator => (operator ? extendedOperators.get(operator)?.form ?? null : null),
-        externalProperty: property,
-      };
-    });
+    const filteringSettings: FilteringSettings = useMemo(() => {
+      const propertyByKey = filteringProperties.reduce((acc, property) => {
+        const extendedOperators = (property?.operators ?? []).reduce(
+          (acc, operator) => (typeof operator === 'object' ? acc.set(operator.operator, operator) : acc),
+          new Map<PropertyFilterOperator, null | ExtendedOperator<any>>()
+        );
+        acc.set(property.key, {
+          propertyKey: property.key,
+          propertyLabel: property?.propertyLabel ?? '',
+          groupValuesLabel: property?.groupValuesLabel ?? '',
+          propertyGroup: property?.group,
+          operators: (property?.operators ?? []).map(op => (typeof op === 'string' ? op : op.operator)),
+          defaultOperator: property?.defaultOperator ?? '=',
+          getValueFormatter: operator => (operator ? extendedOperators.get(operator)?.format ?? null : null),
+          getValueFormRenderer: operator => (operator ? extendedOperators.get(operator)?.form ?? null : null),
+          externalProperty: property,
+        });
+        return acc;
+      }, new Map<string, InternalFilteringProperty>());
+      const getProperty = (propertyKey: string): null | InternalFilteringProperty =>
+        propertyByKey.get(propertyKey) ?? null;
 
-    const propertyByKey = new Map(internalFilteringProperties.map(p => [p.propertyKey, p]));
+      const options: readonly InternalFilteringOption[] = filteringOptions.map(option => {
+        const formatter = getProperty(option.propertyKey)?.getValueFormatter();
+        return {
+          propertyKey: option.propertyKey,
+          value: option.value,
+          label: formatter ? formatter(option.value) : option.label ?? option.value ?? '',
+        };
+      });
+      const optionsByProperty = options.reduce((acc, option) => {
+        const propertyOptions = acc.get(option.propertyKey) ?? [];
+        propertyOptions.push(option);
+        acc.set(option.propertyKey, propertyOptions);
+        return acc;
+      }, new Map<string, InternalFilteringOption[]>());
+      const getPropertyOptions = (propertyKey: string): readonly InternalFilteringOption[] =>
+        optionsByProperty.get(propertyKey) ?? [];
 
-    const internalFilteringOptions: readonly InternalFilteringOption[] = filteringOptions.map(option => {
-      const formatter = propertyByKey.get(option.propertyKey)?.getValueFormatter();
       return {
-        propertyKey: option.propertyKey,
-        value: option.value,
-        label: formatter ? formatter(option.value) : option.label ?? option.value ?? '',
+        properties: [...propertyByKey.values()],
+        options,
+        getProperty,
+        getPropertyOptions,
```

Comment:
2nd system design issue. Components propagate data top down. Propagating getters violates this principle because communication becomes bi directional.


Ideally props propagation should be like a funnel, narrowing down more specific things on each new component level

Reply:
Also see another comment below. Found another red flag in this implementation: https://github.com/cloudscape-design/components/pull/1611#discussion_r1348626190

Reply:
I don't think it violates the React principles as those functions are just getters - an interface on top of the data structure. An alternative could be to pass down structure itself like a bunch of dictionaries and collections, but the getters contribute to better encapsulation.

Consider:
```
function createMapping(...) {
  const propertiesByKey = new Map();
  // ... create mapping from properties
  return propertiesByKey;
}

// and

function createMapping(...) {
  const propertiesByKey = new Map();
  // ... create mapping from properties
  const getProperty = (key: string) => {
    const property = propertiesByKey.get(key);
    if (!key) {
      throw new Error('Invariant violation: missing property');
    }
    return property;
  }
  return { getProperty };
}
```

Both functions perform the same work but the second one results in a better API by validating the output so that consumers should no longer care about the missing properties. It also more restrictive as it does not offer the list of available properties and it can't be mutated.